File: leetcode/Graph/General (Both BFS & DFS)/133. Clone Graph.py
from collections import deque

# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = []):
        self.val = val
        self.neighbors = neighbors

''' 1. DFS '''
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        return self.dfs(node, {})
        
    def dfs(self, node, created):
        if not node:
            return node
        newNode = Node(node.val)
        created[node.val] = newNode
        
        for neighbor in node.neighbors:
            if neighbor.val not in created:
                newNeighbor = self.dfs(neighbor, created)
                newNode.neighbors.append(newNeighbor)
                created[newNeighbor.val] = newNeighbor
            else:
                newNode.neighbors.append(created[neighbor.val])
                
        return newNode

''' 2. BFS '''
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return node
        cloned = {node: Node(node.val)}
        visited = set([node])
        queue = deque([node])
        while queue:
            original = queue.popleft()
            clone = cloned[original]
            for neighbor in original.neighbors:
                if neighbor not in cloned:
                    cloned[neighbor] = Node(neighbor.val)
                clone.neighbors.append(cloned[neighbor])
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        return cloned[node]

File: leetcode/133. Clone Graph.java

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) 
            return node;
        
        Map<Node, Node> cloned = new HashMap<>();
        cloned.put(node, new Node(node.val));

        Set<Node> visited = new HashSet<>();
        visited.add(node);

        Queue<Node> queue = new LinkedList<>();
        queue.offer(node);

        while (!queue.isEmpty()) {
            Node original = queue.poll();
            Node clone = cloned.get(original);
            for (Node neighbor : original.neighbors) {
                if (!cloned.containsKey(neighbor)) {
                    cloned.put(neighbor, new Node(neighbor.val));
                }
                clone.neighbors.add(cloned.get(neighbor));

                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.offer(neighbor);
                }
            }
        }

        return cloned.get(node);
    }
}

// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}